### PR TITLE
Add strict null checks to the codebase

### PR DESCRIPTION
**Pull request by petebacondarwin**
Also converts incorrect use of `void` to `undefined` and adds some words to the CSpell dictionary.

### CODE REVIEW

File: packages/wrangler/pages/functions/filepath-routing.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ export async function generateConfigFromFileTree({
             const declaration = node.declaration;
 
             // `export async function onRequest() {...}`
-            if (declaration.type === "FunctionDeclaration") {
+            if (declaration.type === "FunctionDeclaration" && declaration.id) {
```
Comment (petebacondarwin) [inline packages/wrangler/pages/functions/filepath-routing.ts:55]: Functions that are defined as `export default function` do not have an `id`.
Should we instead add the `default` name to the `exportNames` list?

File: packages/wrangler/pages/functions/filepath-routing.ts [hunk 2]
```diff
@@ -205,7 +203,9 @@ async function forEachFile<T>(
   const returnValues: T[] = [];
 
   while (searchPaths.length) {
-    const cwd = searchPaths.shift();
+    // The `searchPaths.length` check above ensures that `searchPaths.shift()` is defined
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const cwd = searchPaths.shift()!;
```
Comment (petebacondarwin) [inline packages/wrangler/pages/functions/filepath-routing.ts:208]: It is a shame that TS cannot infer this.

File: packages/wrangler/pages/functions/template-worker.ts [hunk 3]
```diff
@@ -110,29 +110,23 @@ export default {
       }
 
       const { value } = handlerIterator.next();
-      if (value) {
```
Comment (petebacondarwin) [inline packages/wrangler/pages/functions/template-worker.ts:113]: There is no need to check the value (according to the types) since it will always be defined.

File: packages/wrangler/src/__tests__/kv.test.ts [hunk 4]
```diff
@@ -212,7 +212,7 @@ describe("wrangler", () => {
 
       it("should make multiple requests for paginated results", async () => {
         // Create a lot of mock namespaces, so that the fetch requests will be paginated
-        const KVNamespaces = [];
+        const KVNamespaces: { title: string; id: string }[] = [];
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/kv.test.ts:215]: Should this be a named type?

File: packages/wrangler/src/inspect.ts [hunk 5]
```diff
@@ -124,8 +124,8 @@ export default function useInspector(props: InspectorProps) {
       server: serverRef.current,
       clientTracking: true,
     });
-    wsServerRef.current.on("connection", (ws: WebSocket) => {
-      if (wsServerRef.current.clients.size > 1) {
+    wsServerRef.current.on("connection", function (ws: WebSocket) {
+      if (this.clients.size > 1) {
```
Comment (petebacondarwin) [inline packages/wrangler/src/inspect.ts:128]: `this` is set to `wsServerRef.current` but not if we use a fat arrow...

File: packages/wrangler/src/dev.tsx [hunk 6]
```diff
@@ -461,25 +461,34 @@ function useEsbuild(props: {
             else {
               // nothing really changes here, so let's increment the id
               // to change the return object's identity
-              setBundle((previousBundle) => ({
-                ...previousBundle,
-                id: previousBundle.id + 1,
-              }));
+              setBundle(
+                (previousBundle) =>
+                  previousBundle && {
```
[the remark this answers is not in the document]
Comment (threepointone) [reply packages/wrangler/src/dev.tsx:466]: I would rather assert that it is defined, and throw if not. (In practice it will never be undefined, because this will be called only after the first build) Alternately, a non-null assertion?

File: packages/wrangler/src/index.tsx [hunk 7]
```diff
@@ -72,10 +72,14 @@ async function readConfig(configPath?: string): Promise<Config> {
     "usage_model",
   ];
 
-  Object.keys(config.env || {}).forEach((env) => {
+  const environments = config.env ?? {};
+  Object.values(environments).forEach((env) => {
+    if (env === undefined) {
+      return;
```
Comment (threepointone) [inline packages/wrangler/src/index.tsx:78]: I don't think we want to return here, instead we'll have to create the object and copy the inherited fields on to it.

File: packages/wrangler/src/index.tsx [hunk 8]
```diff
@@ -86,13 +90,16 @@ async function readConfig(configPath?: string): Promise<Config> {
     "durable_objects",
     "experimental_services",
   ];
-  Object.keys(config.env || {}).forEach((env) => {
+  Object.entries(environments).forEach(([envName, envValue]) => {
+    if (envValue === undefined) {
+      return;
```
Comment (threepointone) [inline packages/wrangler/src/index.tsx:95]: same here, I think we don't want to return early

File: packages/wrangler/src/index.tsx [hunk 9]
```diff
@@ -1233,6 +1249,12 @@ export async function main(argv: string[]): Promise<void> {
               );
             }
 
+            if (args.local) {
```
Comment (threepointone) [inline packages/wrangler/src/index.tsx:1252]: This isn't true, we do support local mode for this command. Did you mean to put this in wrangler tail?

File: packages/wrangler/src/index.tsx [hunk 10]
```diff
@@ -1379,19 +1397,17 @@ export async function main(argv: string[]): Promise<void> {
...
             // -- snip, extract --
 
-            if (!args.local) {
-              const loggedIn = await loginOrRefreshIfRequired();
-              if (!loggedIn) {
-                // didn't login, let's just quit
-                console.log("Did not login, quitting...");
-                return;
-              }
+            const loggedIn = await loginOrRefreshIfRequired();
```
Comment (threepointone) [inline packages/wrangler/src/index.tsx:1400]: why did you remove the args.local checks here?

File: packages/wrangler/src/index.tsx [hunk 11]
```diff
@@ -421,7 +428,11 @@ export async function main(argv: string[]): Promise<void> {
     "👂 Start a local server for developing your worker",
     (yargs) => {
       return yargs
-        .positional("filename", { describe: "entry point", type: "string" })
+        .positional("filename", {
+          describe: "entry point",
+          type: "string",
+          demandOption: true,
```
Comment (threepointone) [inline packages/wrangler/src/index.tsx:434]: using `<filename>` (i.e the pointy brackets) implies `demandOption:true`, but ok.

File: packages/wrangler/src/inspect.ts [hunk 12]
```diff
@@ -148,13 +148,19 @@ export default function useInspector(props: InspectorProps) {
...
    */
   useEffect(() => {
-    serverRef.current.listen(props.port);
-    return () => {
-      serverRef.current.close();
-      // Also disconnect any open websockets/devtools connections
-      wsServerRef.current.clients.forEach((ws) => ws.close());
-      wsServerRef.current.close();
-    };
+    if (serverRef.current) {
```
Comment (threepointone) [inline packages/wrangler/src/inspect.ts:151]: I think we should assert here. It's a big problem if it's not defined, so we should probably throw.

**Review by threepointone — COMMENTED** [review]
Left a first round of comments

File: packages/wrangler/src/dev.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/wrangler/src/dev.tsx:466]: I'll add an error.
But I don't actually understand what the point of this handler is. Is the idea that updating `bundle` by calling `setBundle()` causes the `Dev` component to be re-rendered?

File: packages/wrangler/src/index.tsx (hunk 7 above)
Comment (petebacondarwin) [reply packages/wrangler/src/index.tsx:78]: The previous code would crash if the value of `config.env[key]` was undefined, since it had the following unguarded expression: `config.env[env]` (where in that case `env` is the key.

File: packages/wrangler/src/index.tsx (hunk 8 above)
Comment (petebacondarwin) [reply packages/wrangler/src/index.tsx:95]: Similar to https://github.com/cloudflare/wrangler2/pull/233#discussion_r783082960 - the previous code would crash in this case.

File: packages/wrangler/src/index.tsx (hunk 9 above)
Comment (petebacondarwin) [reply packages/wrangler/src/index.tsx:1252]: Oops! Sorry.

File: packages/wrangler/src/index.tsx (hunk 10 above)
Comment (petebacondarwin) [reply packages/wrangler/src/index.tsx:1400]: Two reasons: 
- they are redundant due to the error being thrown earlier in the function if `args.local` is true
- if this block is wrapped in an if statement then TS is unable to tell that `config.account_id` is not undefined in the `fetchResult()` call below.

File: packages/wrangler/src/index.tsx (hunk 11 above)
Comment (petebacondarwin) [reply packages/wrangler/src/index.tsx:434]: That is true for "runtime" but for type checking the typings are only able to indicate that `args.filename` must not be undefined if the `demandOption` property is true.

File: packages/wrangler/src/inspect.ts (hunk 12 above)
Comment (petebacondarwin) [reply packages/wrangler/src/inspect.ts:151]: An alternative is that we capture these values in a const outside the callback scope...

```ts
      const serverInst = serverRef.current;
      const wsServerInst = wsServerRef.current;
      return () => {
        serverInst.close();
        // Also disconnect any open websockets/devtools connections
        wsServerInst.clients.forEach((ws) => ws.close());
        wsServerInst.close();
      };
```

But I wasn't sure if we could be confident that this would never change between calls to `useEffect()` and its cleanup function...

File: packages/wrangler/src/__tests__/kv.test.ts (hunk 4 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/kv.test.ts:215]: I think it would clarify the API if it was

File: packages/wrangler/src/index.tsx (hunk 11 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/index.tsx:434]: Didn't we recently add the PR from James that allowed for empty `filename` to attempt to run on a root `index`? This would override that. 

https://github.com/cloudflare/wrangler2/pull/79

File: packages/wrangler/pages/functions/filepath-routing.ts (hunk 2 above)
Comment (JacobMGEvans) [reply packages/wrangler/pages/functions/filepath-routing.ts:208]: could include in the `while` check a `&& Array.isArray(searchPaths)` it could be not inferring because `.length` can also be on a string.

Comment (JacobMGEvans) [reply packages/wrangler/pages/functions/filepath-routing.ts:208]: or alternatively 
```suggestion
    const cwd = Array.isArray(searchPaths) && searchPaths.shift()!;
```

Comment (JacobMGEvans) [reply packages/wrangler/pages/functions/filepath-routing.ts:208]: If we run across a lot of these we can make a custom type-guard, this example isn't inferring anything from inputs or using generics.
```typescript
function isNonEmptyArrStrings(value: any): boolean {
    return Array.isArray(value) && value.length && value.every(item => typeof item === "string");
}
```

File: packages/wrangler/pages/functions/routes.ts [hunk 13]
```diff
@@ -114,7 +114,7 @@ export function parseConfig(config: Config, baseDir: string) {
     });
   }
 
-  for (const [route, props] of Object.entries(config.routes)) {
+  for (const [route, props] of Object.entries(config.routes ?? {})) {
```
Comment (JacobMGEvans) [inline packages/wrangler/pages/functions/routes.ts:117]: Wont this break horrifically if it tries to parse an empty object?

File: packages/wrangler/src/index.tsx (hunk 8 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/index.tsx:95]: should it be an error in both cases then?
`throw new Error('Missing ${something} in environments)`

File: packages/wrangler/src/pages.tsx [hunk 14]
```diff
@@ -862,11 +862,13 @@ export const pages: BuilderCallback<unknown, unknown> = (yargs) => {
...
-        // Should only be called if no proxyPort, using `assert.fail()` here
-        // means the type of `assetsFetch` is still `typeof fetch`
-        const assetsFetch = proxyPort
-          ? () => assert.fail()
-          : await generateAssetsFetch(directory);
+        // `assetsFetch()` will only be called if there is `proxyPort` defined.
+        // We only define `proxyPort`, above, when there is no `directory` defined.
+        const assetsFetch =
+          directory !== undefined
+            ? await generateAssetsFetch(directory)
```
Comment (JacobMGEvans) [inline packages/wrangler/src/pages.tsx:869]: seems like we can remove `import assert from "assert";` with current changes

File: packages/wrangler/pages/functions/template-worker.ts (hunk 3 above)
Comment (GregBrimble) [reply packages/wrangler/pages/functions/template-worker.ts:113]: Always present, but not always truthy. Once the iterator is exhausted, `value` will be `undefined`.

File: packages/wrangler/pages/functions/routes.ts (hunk 13 above)
Comment (GregBrimble) [reply packages/wrangler/pages/functions/routes.ts:117]: Why? If `config.routes` is an empty object, then `Object.entries(config.routes) === []`, so the `for` loop will never execute. `routes` will continue to be an empty array, as defined on L71.

File: packages/wrangler/pages/functions/filepath-routing.ts (hunk 1 above)
Comment (GregBrimble) [reply packages/wrangler/pages/functions/filepath-routing.ts:55]: No. Happy to exclude it like you've done.

**Comment by GregBrimble** [comment]
Tried to add a couple of commits, but realized this is on your fork, @petebacondarwin .

So, two tiny change requests:

```diff
--- a/packages/wrangler/src/pages.tsx
+++ b/packages/wrangler/src/pages.tsx
@@ -1,6 +1,5 @@
 /* eslint-disable no-shadow */
 
-import assert from "assert";
 import type { BuilderCallback } from "yargs";
 import { join } from "path";
 import { tmpdir } from "os";
@@ -122,7 +121,9 @@ async function spawnProxyProcess({
       },
     }
   );
-  EXIT_CALLBACKS.push(() => proxy.kill());
+  EXIT_CALLBACKS.push(() => {
+    proxy.kill();
+  });
```

File: packages/wrangler/pages/functions/template-worker.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/wrangler/pages/functions/template-worker.ts:113]: Doh! I should have remembered.

So the "proper" way should be:

```ts
const {value, done} = handleIterator.next();
if (!done) {
 ...
}
```

But then that would mean that we don't handle the final `return` statement from the `executeRequest()` generator.

So I can see that this is the most appropriate approach. Thanks for catching this.
(If only we had unit tests for this...)

**Comment by petebacondarwin** [comment]
@GregBrimble as a "collaborator" on this repo, you have the rights to push to this branch on my fork.

Screenshot 2022-01-13 at 14 57 01

File: packages/wrangler/pages/functions/filepath-routing.ts [hunk 15]
```diff
@@ -155,12 +155,10 @@ export async function generateConfigFromFileTree({
...
 export function compareRoutes(a: string, b: string) {
-  function parseRoutePath(routePath: string) {
-    let [method, segmentedPath] = routePath.split(" ");
-    if (!segmentedPath) {
-      segmentedPath = method;
-      method = null;
-    }
+  function parseRoutePath(routePath: string): [string | null, string[]] {
+    const parts = routePath.split(" ", 2);
+    const segmentedPath = parts.pop() ?? "";
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/wrangler/pages/functions/filepath-routing.ts:160]: Yeah, so it was either this, or adding a `!` and an eslint override comment 🤷

File: packages/wrangler/src/pages.tsx (hunk 14 above)
Comment (petebacondarwin) [reply packages/wrangler/src/pages.tsx:869]: 👍

File: packages/wrangler/src/index.tsx (hunk 11 above)
Comment (petebacondarwin) [reply packages/wrangler/src/index.tsx:434]: The PR you link to was replaced by https://github.com/cloudflare/wrangler2/pull/196/files
Even in this new PR the `filename` option is still required, because that is passed to esbuild as its entry point.

File: packages/wrangler/pages/functions/template-worker.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/wrangler/pages/functions/template-worker.ts:113]: Oh actually, I remember why we needed to remove this.
If we have the `if (value)` then we must also have an `else` clause that returns "something". Otherwise the return type of `next()` becomes `Promise<Response|undefined>` which is not compatible with the `EventContext` type.

File: packages/wrangler/pages/functions/filepath-routing.ts (hunk 2 above)
Comment (petebacondarwin) [reply packages/wrangler/pages/functions/filepath-routing.ts:208]: I'm going to add the guard because that is quite cool

File: packages/wrangler/pages/functions/template-worker.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/wrangler/pages/functions/template-worker.ts:113]: @GregBrimble - can you take a new look at this file now? I have refactored the code a bit.

File: packages/wrangler/src/inspect.ts (hunk 5 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/inspect.ts:128]: that's strange.

File: packages/wrangler/src/index.tsx (hunk 11 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/index.tsx:434]: Hmmm, James had mentioned it defaulting to `index` the 196 PR will be merged soon and we can check the behavior with this change too

**Comment by petebacondarwin** [comment]
Closing and replacing by:

- https://github.com/cloudflare/wrangler2/pull/242
- https://github.com/cloudflare/wrangler2/pull/243
- https://github.com/cloudflare/wrangler2/pull/244
- https://github.com/cloudflare/wrangler2/pull/247
- https://github.com/cloudflare/wrangler2/pull/248
- https://github.com/cloudflare/wrangler2/pull/249
- https://github.com/cloudflare/wrangler2/pull/250
